### ISSUE TITLE
See if we can implement a class type name feature for godot.

### ISSUE THREAD

**Issue by Reousa** [moonwards1/Moonwards-Virtual-Moon#284]

**Comment by yaelatletl** [comment]
Reading a bit on the topic, seems some of the Object functions could be used to get something similar to reflection

**Comment by Reousa** [comment]
> Reading a bit on the topic, seems some of the Object functions could be used to get something similar to reflection
> Edit: Though it would be heavily String based

Yeah I've been reading about this as well, I've been meaning to implement some base classes we'd use for moonwards in GDNative instead, to give us fine-tuned APIs. `MwNode` & `MwSpatial` are examples that would be moved to GDNative at some point, this is also to allow for a better networking implementation.

**Comment by SiudaPL** [comment]
As mentioned above, Godot does have some ways that will tell you which objects you're dealing with.

The [Object](https://docs.godotengine.org/en/stable/classes/class_object.html) class itself has methods like *get_class()*, *is_class(class_name)* or *get_script()*.

Also there is a global object [ClassDB](https://docs.godotengine.org/en/stable/classes/class_classdb.html) which will give you metadata access for every available class. Function here will be like *class_has_method(class, method)* or *get_class_list()*

The override of a [get_class like here](https://godotengine.org/qa/46057/how-to-get-the-class-name-of-a-custom-node) may be helpful for us, since by default it is returning the base class we've inherited, and not the one declared with *class_name* keyword. There is an open [issue](https://github.com/godotengine/godot/issues/21789) regarding that, although hard to say if it's going to be implemented in Godot *4.0*.

One can set it's own metadata with Object: *set_meta(name, value)* and *get_meta(name)*.

Are those methods insufficient for our needs?

**Comment by Reousa** [comment]
As far as I could tell, `ClassDB` isn't exposed to GDScript.

And they are sufficient as preliminary functionality, but the room for error is huge when implemented that way - hence why it's better implemented through GDNative

**Comment by Zach777** [comment]
How detailed are the reflections suppose to be?
 For perfect reflections, we could possible create a second viewport and place it onto the reflecting object.
For imperfect reflections, we could create a shader that loads a pixelized image of what is expected to be reflected on the reflecting body. 

 For bodies that are going to reflect imperfectly and be moving, we could have lightpoints around the world that the body displays on itself when in the correct position.

**Comment by Moonwards** [comment]
The case that is relevant right now is for perfect reflections. I wrote it up as a separate issue, actually - #338

**Comment by Zach777** [comment]
Will the perfect reflections be on moving objects? As reflective moving objects cost a lot of GPU power. Stationary perfect reflective surfaces are a different story.

**Comment by Reousa** [comment]
To confirm, this issue is in relation to GDScript reflection, as in the programming language paradigm, not the reflection of surfaces in the game
